File: src/main/java/com/github/unchama/menuinventory/slot/button/action/ButtonEffectScope.kt
package com.github.unchama.menuinventory.slot.button.action

import com.github.unchama.menuinventory.IndexedSlotLayout
import com.github.unchama.menuinventory.MenuSession
import com.github.unchama.menuinventory.slot.Slot
import com.github.unchama.targetedeffect.UnfocusedEffect
import com.github.unchama.targetedeffect.unfocusedEffect
import org.bukkit.event.inventory.InventoryClickEvent

/**
 * [ButtonEffect]に渡すScoped Lambdaの中で実行可能であるべきメソッドを提供するスコープオブジェクトのクラス.
 */
data class ButtonEffectScope(val event: InventoryClickEvent) {
  fun overwriteCurrentViewBy(newLayout: IndexedSlotLayout): UnfocusedEffect = unfocusedEffect {
    (event.inventory.holder as MenuSession).overwriteViewWith(newLayout)
  }

  fun overwriteCurrentSlotBy(newSlot: Slot): UnfocusedEffect = unfocusedEffect {
    val session = event.inventory.holder as MenuSession
    val newLayout = session.view.slotLayout.altered(event.slot to newSlot)
    session.overwriteViewWith(layout = newLayout)
  }
}

File: src/main/scala/com/github/unchama/menuinventory/slot/button/action/ButtonEffectScope.scala
package com.github.unchama.menuinventory.slot.button.action

import cats.effect.IO
import com.github.unchama.menuinventory.slot.Slot
import com.github.unchama.menuinventory.{LayoutPreparationContext, MenuSession, MenuSlotLayout}
import com.github.unchama.minecraft.actions.MinecraftServerThreadShift
import org.bukkit.event.inventory.InventoryClickEvent

/**
 * [ButtonEffect]に渡すScoped Lambdaの中で実行可能であるべきメソッドを提供するスコープオブジェクトのクラス.
 */
case class ButtonEffectScope(event: InventoryClickEvent) {
  def overwriteCurrentViewBy(newLayout: MenuSlotLayout)
                            (implicit ctx: LayoutPreparationContext,
                             syncShift: MinecraftServerThreadShift[IO]): IO[Unit] =
    event.getInventory.getHolder.asInstanceOf[MenuSession].overwriteViewWith(newLayout)

  def overwriteCurrentSlotBy(newSlot: Slot)
                            (implicit ctx: LayoutPreparationContext,
                             syncShift: MinecraftServerThreadShift[IO]): IO[Unit] = {
    val session = event.getInventory.getHolder.asInstanceOf[MenuSession]

    for {
      oldLayout <- session.currentLayout.get
      newLayout = oldLayout.altered(event.getSlot -> newSlot)
      _ <- session.overwriteViewWith(newLayout)
    } yield ()
  }
}
